text/0052-tx-summary-digest.md

```diff
@@ -0,0 +1,507 @@
+
+When a hardware wallet is asked to sign an MLSAG, we can give it now the `extended_message_digest`
+and steam it the `TxSummary` and the `TxSummaryUnblindingData`, and it can compute the appropriate digest from this for the MLSAG to sign.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+In block version 0, the Ring MLSAG's sign the "extended message" computed roughly as follows:
+
+```mermaid
+graph TB
+A[TxPrefix] -->|merlin| B[[message -32 bytes-]]
+B --> M
+C[pseudo_output_commitments] --> M
+D[range_proofs] --> M
+M[concatenation] --> N
+N[extended_message -many bytes-]
+```
+
+In block version 2, we changed this (in [MCIP 25](0025-confidential-token-ids.md)) so that the Ring MLSAG's sign the
+"extended message digest" which is computed roughly as follows:
+
+```mermaid
+graph TB
+A[TxPrefix] -->|merlin| B[[message -32 bytes-]]
+B --> M
+C[pseudo_output_commitments] --> M
+D[range_proofs] --> M
+M[merlin] --> N
+N[[extended_message_digest -32 bytes-]]
+```
+
+We propose that in block version 3, Ring MLSAGs (other than the [MCIP 31](0031-transactions-with-contingent-inputs.md) Signed Contingent Inputs)
+should now sign the `extended_message_and_tx_summary_digest`,
+which is computed roughly as follows:
+
+```mermaid
+graph TB
+A[TxPrefix] -->|merlin| B[[message -32 bytes-]]
+B --> M
+C[pseudo_output_commitments] --> M
+D[range_proofs] --> M
+M[merlin] --> N
+N[[extended_message_digest -32 bytes-]] --> P
+O[TxSummary] --> P
+P[merlin] --> Q
+Q[[extended_message_and_tx_summary_digest -32 bytes-]]
+```
+
+Example code computing this new digest in a concrete way using the `mc-crypto-digestible` crate:
+
+```rust
+let mut transcript =
+    MerlinTranscript::new(EXTENDED_MESSAGE_AND_TX_SUMMARY_DOMAIN_TAG.as_bytes());
+extended_message.append_to_transcript(b"extended_message", &mut transcript);
+tx_summary.append_to_transcript(b"tx_summary", &mut transcript);
+
+let mut extended_message_and_tx_summary_digest = [0u8; 32];
+transcript.extract_digest(&mut extended_message_and_tx_summary_digest);
+```
+
+This digest is computable given only the 32 byte `extended_message_digest`, and the
+`TxSummary`, which are together much smaller than an entire `Tx`.
+
+The `TxSummary` is a new object with the following schema.
+
+```mermaid
+classDiagram
+direction BT
+TxOutSummary --> TxSummary
+TxInSummary --> TxSummary
+TxSummary: Vec~TxOutSummary~ outputs
+TxSummary: Vec~TxInSummary~ inputs
+TxSummary: uint64 fee_value
+TxSummary: uint64 fee_token_id
+TxSummary: uint64 tombstone_block
+TxSummary: new(&TxPrefix, &[CompressedCommitment])
+TxOutSummary: CompressedRistrettoPublic public_key
+TxOutSummary: CompressedRistrettoPublic target_key
+TxOutSummary: MaskedAmount masked_amount
+TxOutSummary: bool associated_to_input_rules
+TxInSummary: CompressedCommitment pseudo_output_commitment
+TxInSummary: bool has_input_rules
```

Comment:
Ryan made a suggestion that, maybe the `TxInSummary` should include a digest of the input rules rather than just this flag. Then if later, we think it's important to pass all the rules, we can do so without an enclave change. So it's a bit more future proof, and the cost in terms of wire protocol seems very low.